### PR TITLE
User can upload docs for membership application

### PR DESCRIPTION
**Pull request by weedySeaDragon**
PT Story: **User can upload docs for membership application**
https://www.pivotaltracker.com/story/show/133109591

Changes proposed in this pull request:
1.  uses the `paperclip` gem to handle file uploading and connecting a file to a ActiveRecord
2. a **Membership Application** can have many uploaded files; the membership application views now show the ability to upload a file
3. the accepted (valid) content types are images, text,  pdf, and word documents
4. the location for the uploaded files is set only the _TEST_ environment; (DEV uses the defaults; PRODUCTION will be addressed in a different user story)

Users will see a flash message when they have uploaded a file (in addition to a flash message about the membership application).

No flash message shows after a file has been deleted.  (That could be done later if needed.)

ready for review @thesuss !  (finally)

### CODE REVIEW

File: features/upload.feature
Comment (thesuss) [inline features/upload.feature]: Intentation is one too many on this scenario

File: app/controllers/uploaded_files_controller.rb
Comment (tochman) [inline app/controllers/uploaded_files_controller.rb]: This is scaffolded code with a lot of overhead. For example we don't respond to json in other controllers and actions. Why do you do it here?

**Review by tochman — CHANGES_REQUESTED** [review]

File: app/controllers/uploaded_files_controller.rb
Comment (weedySeaDragon) [reply app/controllers/uploaded_files_controller.rb]: Was just using the standard rails g scaffold.  Happy to take the json stuff out.

File: app/controllers/uploaded_files_controller.rb
Comment (tochman) [inline app/controllers/uploaded_files_controller.rb]: Is the show action used in this PR?

File: app/controllers/uploaded_files_controller.rb
Comment (tochman) [inline app/controllers/uploaded_files_controller.rb]: Is the index action used in this PR?

File: app/controllers/uploaded_files_controller.rb
Comment (tochman) [inline app/controllers/uploaded_files_controller.rb]: Is the edit action used in this PR?

File: app/controllers/uploaded_files_controller.rb
Comment (tochman) [inline app/controllers/uploaded_files_controller.rb]: Is the delete action used in this PR?

File: app/controllers/uploaded_files_controller.rb
Comment (tochman) [reply app/controllers/uploaded_files_controller.rb]: It's not just the json stuff. The danger in scaffolding controllers is that you get all sorts of actions (CRUD) that are probably not used.

File: config/routes.rb
Comment (tochman) [inline config/routes.rb]: This creates all CRUD routes. Are they all needed at this point? Also, since the attachments should be related to a specific MembershipApplication, I would suggest making these routes as part or that resource (by nesting them).

**Review by tochman — DISMISSED** [review]

File: app/models/uploaded_file.rb
Comment (patmbolger) [inline app/models/uploaded_file.rb]: I think MIME type `application/msword` is limited to the old ".doc" extension.  

We will need to confirm support for the ".docx" extension as well (I assume).   The MIME type that I have used for that is `application/vnd.openxmlformats-officedocument.wordprocessingml.document` (yup, pretty ugly).

File: app/controllers/membership_applications_controller.rb
Comment (patmbolger) [inline app/controllers/membership_applications_controller.rb:82]: Are we holding off on adding functional (rspec) tests for controllers?  If not, we should think about doing so now, as controller actions get significant functionality.

File: app/controllers/uploaded_files_controller.rb
Comment (patmbolger) [inline app/controllers/uploaded_files_controller.rb]: Are we holding off on adding functional (rspec) tests for controllers?  If not, we should think about doing so now, as controller actions get significant functionality.

File: app/models/membership_application.rb
Comment (patmbolger) [inline app/models/membership_application.rb:5]: Consider adding unit test(s) (rspec model) for this.

File: spec/models/uploaded_file_spec.rb
Comment (patmbolger) [inline spec/models/uploaded_file_spec.rb:53]: There is no type for ".doc" files above.

Also, probably need to add support for ".docx" types.

File: spec/models/uploaded_file_spec.rb
Comment (patmbolger) [inline spec/models/uploaded_file_spec.rb:79]: Consider adding tests for the `#save` action, and confirm correct behavior.

That is, success == UploadedFiles.count increases by 1.

failure = correct error message(s) (e.g., invalid file type, file not found, etc.)

File: spec/factories/uploaded_files.rb
Comment (patmbolger) [inline spec/factories/uploaded_files.rb:43]: Missing .doc and .docx file types.  (I assume these types are supported?  - if not, please ignore)

**Review by patmbolger — COMMENTED** [review]
Reviewed .... please see comments.

File: app/controllers/uploaded_files_controller.rb
Comment (tochman) [reply app/controllers/uploaded_files_controller.rb]: I mentioned this before but I see that a lot of scaffolded code is still here. All actions that are not needed should be removed from the scaffold. Comments as well.

File: app/models/uploaded_file.rb
Comment (weedySeaDragon) [reply app/models/uploaded_file.rb]: Argh -- I'd seen that ugly MIME type and was hoping it wasn't really needed.  I'm glad to learn of your experience: that _is_ the MIME type for the .docx document type.

File: app/controllers/uploaded_files_controller.rb
Comment (weedySeaDragon) [reply app/controllers/uploaded_files_controller.rb]: @tochman -- I'm happy to remove stuff.  I'm finding it hard to know what scaffolded stuff to leave in and what to take out.  Convention = leave it in because most projects need it.  That's why it's become the conventional standard and included in the scaffolds. 

If the standard on this project should be to not include so much of it, then I'd say that the generators & templates should be adjusted accordingly.  That way no one can run into this problem because the scaffolds will create exactly what is needed and no more. 

My sense is that there is some subset of scaffolding that is emphasized in CraftAcademy, and so that is the implicit guideline for this project. But for those of us that weren't in CraftAcademy, in the absence of any other information we can only assume that what is generated in the scaffolded code is acceptable.

Comment (tochman) [reply app/controllers/uploaded_files_controller.rb]: IMO a PR should only include the code it needs to deliver the requested functionality. Do you see it differently?

**Comment by tochman** [comment]
I can't see that an index and show views for uploads are part of this feature.  I also feel that the destroy feature is out of scope. But I'll leave this to somebody else to comment upon. IMO this PR is way to big to grasp and I feel I need to excuse myself from the review process.

File: app/controllers/uploaded_files_controller.rb
Comment (weedySeaDragon) [reply app/controllers/uploaded_files_controller.rb]: No, of course not.  But that guideline doesn't exist in isolation.  That assumes that the user story is more rigorous than is the reality of this project.  IMO, there has not been the time or process or conversations necessary to develop the business goals and user stories.  (I don't know if that's because this is in part a learning project, or if that's a deliberate decision or a combination of those two.)

Conventions over configuration isn't just about saving time.  It also speaks to the the fact that the scaffold code is there because most projects have found it necessarily and valid.  And using code that is mature and well vetted reduces risk and saves time. 

In this particular project, I think the costs and risks of including code produced by the very standard rails scaffolding are very, very low. It is a small enough set of code that it doesn't confuse or impact readability.  It is functionality that most projects need (even if not at their very first iteration). If some particular performance testing points to serious gains by removing the code, then it should absolutely be culled.

I still totally support revising the generators and templates to whatever is appropriate for this project.  Otherwise this will simply be an ongoing issue for code reviews, as developers who come in to this project are going to use standard rails tools, which includes the generators.  To me that may be the best thing to do:  This is the 2nd signal that there's a problem in our process, so let's find and fix the root of the problem.

File: app/controllers/membership_applications_controller.rb
Comment (weedySeaDragon) [reply app/controllers/membership_applications_controller.rb:82]: Agreed.  Should there be a chore for making more complete rspec tests for the MembershipApplicationController?

Comment (patmbolger) [reply app/controllers/membership_applications_controller.rb:82]: I guess that depends on project philosophy.  Some projects might require that a "slice" of functionality (such us upload membership files) include *all* elements of the complete stack - including unit, functional and acceptance tests.  Another philosophy is to break up such a "slice" into multiple stories, for different aspects of the stack.  

Either way works.  

The first approach does lead to bigger PR's - and this can be a negative if the architecture (and even the user story) has some ambiguity - which means smaller PR's are safer since less work goes into something that might change significantly with PR reviews.

Any other thoughts out there?

File: app/controllers/uploaded_files_controller.rb
Comment (tochman) [reply app/controllers/uploaded_files_controller.rb]: Rails generators are one thing. A scaffold is, imo, something else.

File: app/controllers/membership_applications_controller.rb
Comment (tochman) [reply app/controllers/membership_applications_controller.rb:82]: @patmbolger Until now we have used Acceptance tests (using Cucumber) and Unit tests (using RSpec only) and stayed away from Controller specs. Imo, the very nature of acceptance tests (running through the entire stack) implicitly tests the controller actions (restful or not) and that is often enough.

Comment (patmbolger) [reply app/controllers/membership_applications_controller.rb:82]: @tochman - ok, makes sense.  Will stick to unit and acceptance tests.

**Comment by weedySeaDragon** [comment]
Recent push = handled merge conflicts.

(now I'll resume working on this to get other issues addressed)

File: app/views/uploaded_files/index.html.haml
Comment (thesuss) [reply app/views/uploaded_files/index.html.haml]: If this line is modern maybe it's not too long? 
`method: :delete`, etc

File: app/views/uploaded_files/_form.html.haml
Comment (thesuss) [reply app/views/uploaded_files/_form.html.haml]: = and " unneccessary

File: app/views/membership_applications/_uploaded_files_list.html.haml
Comment (thesuss) [reply app/views/membership_applications/_uploaded_files_list.html.haml]: could `%td{class: "delete-uploaded-file-#{uploaded_file.id}"}`
be rewritten as `%td.delete-uploaded-file-#{uploaded_file.id}` or maybe
`%td.delete-uploaded-file-=uploaded_file.id`

File: app/views/uploaded_files/index.html.haml
Comment (weedySeaDragon) [reply app/views/uploaded_files/index.html.haml]: It used to be that you had to use the old hashrocket with that (passing the delete method and data in).   That may have changed; maybe it can be written in the newer style.  (someone needs to check and can change this)

Comment (thesuss) [reply app/views/uploaded_files/index.html.haml]: Added a ticket for that ;)

File: app/views/uploaded_files/_form.html.haml
Comment (thesuss) [reply app/views/uploaded_files/_form.html.haml]: We don't even need to count the errors. =)

**Comment by weedySeaDragon** [comment]
Need to determine if we do need the `uploaded_files_controller` file.  It is really required? (since no user is working with uploaded_files independently from a membership application)
 will all tests pass without it? (they might)

File: app/views/membership_applications/_uploaded_files_list.html.haml
Comment (thesuss) [reply app/views/membership_applications/_uploaded_files_list.html.haml:17]: Ach, shut up :)

**Review by thesuss — APPROVED** [review]

**Pull request merged**
